Modifications for IND-CCA-secure DHKEM and independent random oracles (#50)
cfrg/draft-irtf-cfrg-hpke: draft-irtf-cfrg-hpke.md

    of the private key `skR` is assured that the ephemeral shared
    key is known only to the holder of the private key corresponding
    to `pkS`
  - Nzz: The length in bytes of a shared secret produced by this KEM
  - Nenc: The length in bytes of an encapsulated key produced by this KEM
  - Npk: The length in bytes of an encoded public key for this KEM

* A Key Derivation Function (KDF):

>   - Hash(m): Compute the cryptographic hash of input message `m`

Reviewer: Can we get rid of `Hash()` now?

Author: I also had this thought but didn't follow it yet, thanks for bringing it up. I think we can get rid of Hash. I'll have a look where in the spec this would need adjustment.

Author: Removed Hash (and Nb) with commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/877237816152d2a58dc866bba1c9e12b3c4a5a93.

## DH-Based KEM

Suppose we are given a Diffie-Hellman group, and a hash function,
providing the following operations:

- GenerateKeyPair(): Generate an ephemeral key pair `(sk, pk)`
  for the DH group in use
- DH(sk, pk): Perform a non-interactive DH exchange using the
  private key sk and public key pk to produce a Diffie-Hellman

>   shared secret of length Npk

Reviewer: It seems like it might be worth defining a different length, say `Ndh` for this, since there are definitely schemes where the shared secret is not of length `Npk`.

- Marshal(pk): Produce a fixed-length byte string
  encoding the public key `pk`
- Unmarshal(enc): Parse a fixed-length byte string to recover a
  public key
- HashDH(m): Compute the cryptographic hash of input message `m`
  producing the shared secret of length Nzz returned by the algorithm

> - label_hashdh: The UTF-8 string literal "RFCXXXX DHKEM" of length

Reviewer: I wonder if it would be simplest to do something like the following

```
def LabeledExtract(salt, label, secret):
    labeledSecret = concat("RFC XXXX", label, secret)  
    return Extract(salt, labeledSecret)
```

This is what TLS 1.3 does to make sure the general label gets added everywhere.

Reviewer: Is there a reason to attach the label to the secret vs. the salt?

Author: > Is there a reason to attach the label to the secret vs. the salt?

The goal of the label in the end is to separate the input domains of all HMAC calls. This concerns Extract and Expand. The three Expand calls for key, nonce, exporter_secret already have the label as prefix of the second argument, and that's why it's in the second argument for Extract, too (and thus attached to the secret/IKM).

Author: > I wonder if it would be simplest to do something like the following […]
> This is what TLS 1.3 does to make sure the general label gets added everywhere.

Ups, I forgot that in my recent update. I'll first write some comments to what I did there, and later try to incorporate this suggestions.

- GenerateKeyPair(): Generate an ephemeral key pair `(sk, pk)`
  for the DH group in use
- DH(sk, pk): Perform a non-interactive DH exchange using the
  private key sk and public key pk to produce a Diffie-Hellman
  shared secret of length Npk
- Marshal(pk): Produce a fixed-length byte string
  encoding the public key `pk`
- Unmarshal(enc): Parse a fixed-length byte string to recover a
  public key

And consider the following additional definition:

- Hash'(m): Compute the cryptographic hash of input message `m`

>   producing the shared secret of length Nzz returned by the algorithm

Reviewer: `Nzz` needs to be defined somewhere; TBH seems like this would be good to know for other KEMs as well.  I think this loses the idea that the DH output is fixed-size, though.  Do we need that?

Reviewer 2: Putting it alongside the `Nh` et al. definitions would suffice. 

> I think this loses the idea that the DH output is fixed-size, though. Do we need that?

@bifurcation I'm not sure I follow. Nzz *is* fixed, so isn't the output fixed size?

Author: Good point, I'll add `Nzz` to the “Cryptographic Dependencies” part; I am not sure what you mean regarding the fixed-size DH output?

Author: Might be relevant: Encap and AuthEncap now have the same output size for `zz`. Before, AuthEncap's `zz` had double the size of Encap's `zz`.

Reviewer 2: Interesting -- does that affect analysis?

Author: Not so much I'd say – it makes analysis of Extract a bit easier because it's input length has one case distinction less.

primitives is provided in {{ciphersuites}}.  Algorithm identifier
values are two bytes long.

The following two functions are defined for a KDF to facilitate domain
separation of calls as well as context binding:

~~~

> def LabeledExtract(salt, label, IKM):

Reviewer: You use `0` for the salt value several times below.  It would be good to clarify whether that means (a) an all-zero octet string (and if so, what length), or (b) the empty octet string.

Author: In commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/e7313ea2845c72a543dcc32f1eb2be99acb2d9d9, I chose to make explicit what HKDF-Extract would do by default, by using `zero(Nh)`.

>   labeled_IKM = concat("RFCXXXX ", label, IKM)

Reviewer: Nit: I would remove the `_` here.

>   return Extract(salt, labeled_IKM)

Reviewer: Is there a reason to put the label on the IKM vs. the salt?

Author: In the end, the goal of the label is to separate the input domains of all HMAC calls. This concerns both Extract and Expand. The three Expand calls for key, nonce, exporter_secret already have a label as prefix of the second argument, and that's why it's in the second argument for Extract, too (and thus attached to the IKM).

  private key sk and public key pk to produce a Diffie-Hellman
  shared secret of length Ndh
- Marshal(pk): Produce a byte string of length Npk
  encoding the public key `pk`
- Unmarshal(enc): Parse a byte string of length Npk to recover a
  public key
- Ndh: The length in bytes of a Diffie-Hellman shared secret produced
  by the DH function of this KEM.

Then we can construct a KEM called `DHKEM(Group, KDF)` in the
following way, where `Group` denotes the Diffie-Hellman group and
`KDF` the KDF:

~~~
def ExtractAndExpand(dh, context_kem):
  prk = LabeledExtract(0, "dh", dh)
  return LabeledExpand(prk, "prk", context_kem, Nzz)

def Encap(pkR):
  skE, pkE = GenerateKeyPair()
  dh = DH(skE, pkR)
  enc = Marshal(pkE)

  pkRm = Marshal(pkR)
  context_kem = concat(enc, pkRm)

  zz = ExtractAndExpand(dh, context_kem)
  return zz, enc

def Decap(enc, skR, pkR):
  pkE = Unmarshal(enc)
  dh = DH(skR, pkE)

  pkRm = Marshal(pkR)
  context_kem = concat(enc, pkRm)

  zz = ExtractAndExpand(dh, context_kem)
  return zz, enc

def AuthEncap(pkR, skS, pkS):
  skE, pkE = GenerateKeyPair()
  dh = concat(DH(skE, pkR), DH(skS, pkR))
  enc = Marshal(pkE)

  pkRm = Marshal(pkR)
  pkSm = Marshal(pkS)

>   context_kem = concat(enc, pkRm, pkSm)

Reviewer: Nit: Suggest `kemContext`

  zz = ExtractAndExpand(dh, context_kem)
  return zz, enc

def AuthDecap(enc, skR, pkR, pkS):
  pkE = Unmarshal(enc)
  dh = concat(DH(skR, pkE), DH(skR, pkS))

  pkRm = Marshal(pkR)
  pkSm = Marshal(pkS)
  context_kem = concat(enc, pkRm, pkSm)

  zz = ExtractAndExpand(dh, context_kem)
  return zz, enc
~~~

The KDF used in DHKEM can be equal to or different from the KDF used
in the remainder of HPKE depending on the chosen variant. See

> {{domain-separation}} for a security-related discussion.

Reviewer: Do we have guidance on what the right answer is?

Author: In this pull request, I suggest one particular KDF for each of DHKEM's DH groups. This recommendation is based on the security level of the DH group. We could include as guidance that choosing the same KDF for the remainder of KDF is the best trade-of: this keeps the security level of DHKEM while not pretending a higher security level by using a KDF with greater output length.

Does this go into the direction you meant to go with your question?

## DH-Based KEM

Suppose we are given an instantiation KDF_kem of KDF with functions
Hash_kem, Extract_kem, and Expand_kem, and a Diffie-Hellman group
providing the following operations:

- GenerateKeyPair(): Generate an ephemeral key pair `(sk, pk)`
  for the DH group in use
- DH(sk, pk): Perform a non-interactive DH exchange using the
  private key sk and public key pk to produce a Diffie-Hellman

>   shared secret of length Ndh

Reviewer 2: Can we define `Ndh` on its own line? (I missed it the first time around.) Is it even needed?

Author: This was a suggestion by @bifurcation, with the background that there might be a DH scheme where the result of the DH operation is not an encoded public key.

We can move it to it's own line I think.

- Marshal(pk): Produce a byte string of length Npk
  encoding the public key `pk`
- Unmarshal(enc): Parse a byte string of length Npk to recover a
  public key

> - Nzz: The output size of the Hash_kem and Extract_kem functions

Reviewer 2: Re-defining this is somewhat confusing -- is it needed?

Author: We also redefine GenerateKeyPair, DH, Marshal, Unmarshal to make their definitions adapted to DHKEM. With the line about Nzz I want to make clear that Nzz is the length of Hash_kem's output. This is only implicitly clear by seeing that the result of Expand is returned.

It's true that this (Nzz is the length of Hash_kem's output) is true for our variants of DHKEM with SHA2. But now that I included Nzz as third argument to Expand, other settings are also possible.

So maybe we could move this line away from the definitions list and put it as a comment below the pseudo-code, saying that for our specified versions of DHKEM, Nzz is Hash_kem's output length. What do you think about that?

Reviewer 2: > So maybe we could move this line away from the definitions list and put it as a comment below the pseudo-code, saying that for our specified versions of DHKEM, Nzz is Hash_kem's output length. What do you think about that?

That'd be great!

  in bytes

Then we can construct a KEM called `DHKEM(Curve, Hash_kem)` in the
following way, where `Curve` denotes the Diffie-Hellman group and

> `Hash_kem` the hash function underlying KDF_kem:

Reviewer 2: ```suggestion
Then we can construct a KEM called `DHKEM(Group, Hash_kem)` in the
following way, where `Group` denotes the Diffie-Hellman group and
`Hash_kem` the hash function underlying KDF_kem:
```

~~~
label_dh = "RFCXXXX DHKEM"

def Encap(pkR):
  skE, pkE = GenerateKeyPair()
  dh = DH(skE, pkR)
  enc = Marshal(pkE)

  pkRm = Marshal(pkR)
  context_kem = concat(enc, pkRm)

>   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))

Reviewer 2: ```suggestion
  prk = Extract_kem(0, concat(label_dh, dh))
```

>   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))

Reviewer 2: These lines (prk = ..., zz = ...) are the same across all variants. Can we pull them out into a common function that takes `dh`, `context_kem` as input?

```
def ExtractAndExpand(dh, prk):
   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))
   zz  = Expand_kem(prk, concat("prk", context_kem), Nzz)
```

Then call like so:

```
def Encap(pkR):
   ...
   zz = ExtractAndExpand(dh, context_kem)
   return zz, enc
```

  zz  = Expand_kem(prk, concat("prk", context_kem), Nzz)
  return zz, enc

def Decap(enc, skR, pkR):
  pkE = Unmarshal(enc)
  dh = DH(skR, pkE)

  pkRm = Marshal(pkR)
  context_kem = concat(enc, pkRm)

>   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))

Reviewer 2: ```suggestion
  prk = Extract_kem(0, concat(label_dh, dh))
```

  zz  = Expand_kem(prk, concat("prk", context_kem), Nzz)
  return zz, enc

def AuthEncap(pkR, skS, pkS):
  skE, pkE = GenerateKeyPair()
  dh = concat(DH(skE, pkR), DH(skS, pkR))
  enc = Marshal(pkE)

  pkRm = Marshal(pkR)
  pkSm = Marshal(pkS)
  context_kem = concat(enc, pkRm, pkSm)

>   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))

Reviewer 2: ```suggestion
  prk = Extract_kem(0, concat(label_dh, dh))
```

  zz  = Expand_kem(prk, concat("prk", context_kem), Nzz)
  return zz, enc

def AuthDecap(enc, skR, pkR, pkS):
  pkE = Unmarshal(enc)
  dh = concat(DH(skR, pkE), DH(skR, pkS))

  pkRm = Marshal(pkR)
  pkSm = Marshal(pkS)
  context_kem = concat(enc, pkRm, pkSm)

>   prk = Extract_kem(0, concat("RFCXXXX DHKEM", dh))

Reviewer 2: ```suggestion
  prk = Extract_kem(0, concat(label_dh, dh))
```

  VerifyMode(mode, psk, pskID, pkSm)

  pkRm = Marshal(pkR)
  ciphersuite = concat(encode_big_endian(kem_id, 2),
                       encode_big_endian(kdf_id, 2),
                       encode_big_endian(aead_id, 2))
  pskID_hash = Extract(0, concat("pskID", pskID))
  info_hash = Extract(0, concat("info", info))
  identifier = "RFCXXXX HPKE"
  context = concat(identifier, ciphersuite, mode, enc, pkRm,
                   pkSm, pskID_hash, info_hash)

  if length(psk) > Nb then:
    psk = Extract(0, concat("psk", psk))

  secret = Extract(psk, concat("zz", zz))
  key = Expand(secret, concat("key", context), Nk)
  nonce = Expand(secret, concat("nonce", context), Nn)
  exporter_secret = Expand(secret, concat("exp", context), Nh)

  return Context(key, nonce, exporter_secret)
~~~~~
\[\[RFC editor: please change "RFCXXXX" to the correct number before publication.]]

Note that the context construction in the KeySchedule procedure is
equivalent to serializing a structure of the following form in the
TLS presentation syntax:

~~~~~
struct {

>     uint8 identifier[12];

Reviewer 2: ```suggestion
    uint8 identifier[12] = "RFCXXXX HPKE";
```

> 

Author: Is this correct for the 12-letter UTF-8 string literal “RFCXXXX HPKE”?

Reviewer 2: Let's treat it as an ASCII string, in which case, yes!

  VerifyMode(mode, psk, pskID, pkSm)

  pkRm = Marshal(pkR)
  ciphersuite = concat(encode_big_endian(kem_id, 2),
                       encode_big_endian(kdf_id, 2),
                       encode_big_endian(aead_id, 2))
  pskID_hash = LabeledExtract(0, "pskID", pskID)
  info_hash = LabeledExtract(0, "info", info)
  context = concat(ciphersuite, mode, enc, pkRm,
                   pkSm, pskID_hash, info_hash)

>   if len(psk) > Nb then:

Author: Is it ok to use the already defined `len()` for the psk?
```
- `len(x)`: The length of the byte string `x`, expressed as a
  two-byte unsigned integer in network (big-endian) byte order
```
For now `len()` is only used to get the length of the nonce:
```
def Context.Nonce(seq):
  encSeq = encode_big_endian(seq, len(self.nonce))
  return xor(self.nonce, encSeq)
```
By the way, why compute the length of the nonce here? We know that it has length `Nn`. Also, why does encode_big_endian apparently need the length in “network (big-endian) byte order” and as two-byte integer?

Reviewer 2: Yep, it's fine, and preferred! We pass a length in case the resulting number needs to be padded with zeros.

  VerifyMode(mode, psk, pskID, pkSm)

  pkRm = Marshal(pkR)
  ciphersuite = concat(encode_big_endian(kem_id, 2),
                       encode_big_endian(kdf_id, 2),
                       encode_big_endian(aead_id, 2))
  pskID_hash = LabeledExtract(0, "pskID", pskID)
  info_hash = LabeledExtract(0, "info", info)
  context = concat(ciphersuite, mode, enc, pkRm,
                   pkSm, pskID_hash, info_hash)

>   if len(psk) > Nb then:

Reviewer: Why do we need this explicit hashing step?  Shouldn't the internals of `LabeledExtract` handle that for us?

Author: Are you suggesting this if should be inside LabeledExtract? This is the only place where we give an arbitrary-length bitstring as the first parameter to (Labeled)Extract, so maybe it's wasteful to have the branching every time (even for all the calls with 0 as first parameter)? If you say you prefer this to have a cleaner KeySchedule, I think I could agree.

Just in case: Have you seen my earlier comment? It might have been buried within this long discussion, so let me cite:

https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50#issuecomment-603568796

> On the PSK length:
> 
> HMAC(K, M) feeds K through a call to Hash if K is longer than Hash's blocksize. If we allow arbitrary-size PSKs for user convenience, this would introduce another place where Hash is called on arbitrary-sized input, just like with info and pskID. (As a reminder: Arbitrary-sized input to Hash can lead to collisions with input to Hash calls inside HMAC. This would render the proof inapplicable that proves indifferentiability of HMAC from a random oracle. To avoid collisions, we use Extract calls and separate the input domains by using a prefix).
> 
> To avoid this extra Hash call inside HMAC, we can just anticipate it directly in KeySchedule, and thus control the prefix. While this introduces branching inside KeySchedule, this branching would either way be done inside HMAC. Doing the Extract call even if the PSK is shorter then Nb would hurt performance in non-PSK modes, too.

Author: I am reading specs of Keccak and Blake2/3 to see if they do internal hashing in their KDF/XOF modes that could collide with a Hash call inside HPKE. This `len > Nb` step is very specific to HMAC, and people might implement Extract and Expand with something different. So maybe we can come up with a more general statement here instead of hardcoding support for HMAC.

Reviewer: Right, I was thinking that in the HKDF case, this duplicated what HMAC was doing internally.  If there's nothing current that is doing something that would collide, I would suggest we remove this, and maybe add a note about the assumption we're making about the KDF (no internal hashing that could collide).

Reviewer 2: @bifurcation this isn't quite duplicating what HMAC is doing internally, since it explicitly prefixes the input before hashing.

Author: With commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/421828db601168b9f81acc6d299754e477e4d10a, the PSK is now always fed through LabeledExtract. I included my reasoning in the commit message. Please let me know if that makes sense to you.

Author: An alternative to feeding the PSK always through LabeledExtract is to restrict the PSK to a fixed length, for example to Nh.

Reviewer 2: 👍

Author: If someone implements Extract and Expand with something else than HKDF, like a modern hash function's KDF and XOF modes, then this PSK hashing step is probably not needed. I attempted a formulation with commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/09acc770c7604185836085499b7284dcfa0568a8.

* The PSK ID is added to the context string used as the `info` input
  to the KDF

The PSK SHOULD be of length Nh bytes or longer, and SHOULD have
Nh bytes of entropy or more. Using a PSK shorter than Nh bytes

> is permitted. A PSK that is longer than Nh bytes or has more than

Reviewer 2: Can we pull the text from #59 and place it here? (And move the next paragraph to the security considerations as is done in #59?)

## Key Encapsulation Mechanisms (KEMs) {#kem-ids}

| Value  | KEM                       | Nzz  | Nenc | Npk | Reference                    |
|:-------|:--------------------------|:-----|:-----|:----|:-----------------------------|
| 0x0000 | (reserved)                | N/A  | N/A  | N/A | N/A                          |

> | 0x0010 | DHKEM(P-256, SHA256)      | 32   | 65   | 65  | {{NISTCurves}}, {{?RFC6234}} |

Reviewer 2: Can we introduce this notation (`DHKEM(curve, hash)`) in the DHKEM section?

conditions:

- DH group: The gap Diffie-Hellman (GDH) problem is hard {{GAP}}.
- Extract and Extract_kem: Indifferentiable from a random oracle,

>   respectively.

Reviewer: What does "respectively" mean here?  If you mean that they are each individually indifferentiable, I would phrase as "Each function is individually indifferentiable from a random oracle".

Reviewer: I assume this bullet and the next one are supposed to be "KDF used in DHKEM" and "KDF used in the rest of HPKE"?  If so, let's just say that.  Something like:

```
- Extract and Expand (in DHKEM): ...
- Extract and Expand (elsewhere): ...
```

Author: Ups, sorry that one slipped through after removing the `_kem` suffix. Yes, it means “Each function […]”.

Author: Addressed in commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/fbea8c041a50b6db3a458b7c2cdefa4cdb1014e8.

computation. A full proof of post-quantum security would need to take this
difference into account, in addition to simply using a post-quantum KEM.

> ## Random Oracle Cloning / Domain Separation

Reviewer 2: ```suggestion
## Domain Separation
```

... since Random Oracle Cloning isn't a widely used term yet.

HPKE allows combining a DHKEM variant DHKEM(Curve, Hash_kem) and a KDF
such that the hash functions Hash_kem and Hash are instantiated by the
same hash function. The prefixes used for the second parameter of the
functions Extract_kem, Expand_kem, Extract, and Expand serve to separate
their input domains; this justifies modeling them as independent
functions even if instantiated by the same underlying hash function.

Users of HPKE who want to use a KEM other than DHKEM must make sure
that any hash function used within this KEM can be modeled as a

> function independent from `Hash`.

Reviewer 2: ```suggestion
Future KEM instantiations MUST ensure that all internal hash function invocations are prefixed with a unique label similar to `"RFCXXXX DHKEM"`. This ensures they can be modeled as a function independent from `Hash`.
```

computation. A full proof of post-quantum security would need to take this
difference into account, in addition to simply using a post-quantum KEM.

## Pre-Shared Key Recommendations {#security-psk}

In mode PSK and mode AuthPSK, the PSK SHOULD be of length Nh bytes or
longer, and SHOULD have Nh bytes of entropy or more. Using a PSK shorter
than Nh bytes is permitted. A PSK that is longer than Nh bytes or that
has more than Nh bytes of entropy, respectively, does not increase the
security level of HPKE, because the extraction step involving the PSK
only outputs Nh bytes.

HPKE is specified to use HKDF as key derivation function. HKDF is not
designed to slow down dictionary attacks, see {{?RFC5869}}. Thus,
HPKE's PSK mechanism is not suitable for use with a low-entropy
password as the PSK: in scenarios in which the adversary knows the
KEM shared secret zz, it can perform a dictionary attack on the PSK.

> The adversary can use different oracles to decide if a guess was good,

Reviewer 2: ```suggestion
The adversary can use different oracles to decide if a PSK guess was correct,
```

> depending on the scenario. First, on observation of an HPKE ciphertext,

Reviewer 2: In this scenario, the adversary didn't guess the PSK value, so I don't think it's a relevant example.

Author: The adversary can try decryption for each PSK value it wants to guess. This is an example of a passive adversary. Would replacing “on observation of an HPKE ciphertext” by “with a captured HPKE ciphertext” make this clearer?

Reviewer 2: If the adversary doesn't have access to the ephemeral shares, this isn't possible, right?

> it can use decryption of the ciphertext as oracle. HPKE ciphertexts are

Reviewer 2: ```suggestion
the adversary can use decryption of the ciphertext as oracle. HPKE ciphertexts are
```

created with an authenticated encryption scheme, which means decryption
fails when using a wrong key. Second, the adversary can act as sender
and create candidate HPKE ciphertexts. A recipient's behaviour can serve
as oracle, e.g. if an error message implies that a wrong PSK was used.
Third, if an application does not use Seal to create HPKE ciphertexts
but only the Export interface, the adversary could rely on other
recipient behavior which is observably different when using a wrong key.

> Scenarios in which the adversary can know the KEM shared secret zz

Reviewer 2: ```suggestion
Scenarios in which the adversary knows the KEM shared secret zz
```

depend on the KEM. In the case of DHKEM, it is enough if the adversary
knows all private keys of one participant. In mode PSK, either knowledge
of skR or skE is sufficient. If the adversary can act as sender, it does
obviously know skE. In mode AuthPSK, either knowledge of skR or (skS and
skE) is sufficient. Knowledge of skR is sufficient because

> Diffie-Hellman is subject to key-compromise impersonation.

Reviewer 2: ```suggestion
knows all private keys of one participant. 
```

I recommend deleting these examples as knowledge of all private keys of one participant is adequate.

computation. A full proof of post-quantum security would need to take this
difference into account, in addition to simply using a post-quantum KEM.

## Pre-Shared Key Recommendations {#security-psk}

> In mode PSK and mode AuthPSK, the PSK SHOULD be of length Nh bytes or

Reviewer: "In the PSK and AuthPSK modes, ..."

longer, and SHOULD have Nh bytes of entropy or more. Using a PSK shorter
than Nh bytes is permitted. A PSK that is longer than Nh bytes or that
has more than Nh bytes of entropy, respectively, does not increase the
security level of HPKE, because the extraction step involving the PSK
only outputs Nh bytes.

HPKE is specified to use HKDF as key derivation function. HKDF is not
designed to slow down dictionary attacks, see {{?RFC5869}}. Thus,
HPKE's PSK mechanism is not suitable for use with a low-entropy
password as the PSK: in scenarios in which the adversary knows the
KEM shared secret zz and has access to an oracle for decryption success
of an HPKE ciphertext, it can perform a dictionary attack on the PSK.

> If an application does not use Seal to create HPKE ciphertexts but only

Reviewer: How does Seal/Open prevent this?  If you behave differently on decryption failure, then it seems like you still create an oracle.

Author: New proposal:

Thus, HPKE's PSK mechanism is not suitable for use with a low-entropy
password as the PSK: in scenarios in which the adversary knows the
KEM shared secret zz and has access to an oracle that allows to distinguish
between a good and a wrong key, it can perform a dictionary attack on the PSK.
This oracle can be the decryption operation on a captured HPKE ciphertext or
any other recipient behavior which is observably different when using a wrong key.

Reviewer: Works for me, thanks.

the Export interface, the adversary can rely on other recipient behavior
which is observably different when using a wrong key.

Scenarios in which the adversary knows the KEM shared secret zz
depend on the KEM. In the case of DHKEM, it is enough if the adversary

> knows all private keys of one participant. 

Reviewer: What do you mean by "it is enough"?  Also, I don't understand the dependency on DHKEM here.

Author: The adversary knows the KEM shared secret zz in certain compromise scenarios. We were hesitant to make a general claim about compromise scenarios with a general KEM. But maybe it's actually possible. Have to think about it.

The “enough” because people could think that for DHKEM's AuthPSK it is not enough to know only the recipient's private static key, but that knowing the sender's private static key is also necessary; because of key-compromise impersonation this is not the case. We could do without the “enough” and say “For DHKEM this is the case if the adversary knows all private keys of one participant”.

Author: For a “good or bad PSK” oracle, knowing the secrets of _one_ party is enough because: Each one of the two parties involved in HPKE must be able to decrypt a message, and each party only knows its own secrets.

I will rephrase the paragraph accordingly.

Reviewer: I see.  I think I misunderstood the "Scenaris where the adversary knows zz" above.  With your new text above, it's clearer.

Author: To everyone: I attempted to clarify this in commit https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/50/commits/1c84680727de94679dcf1f96c7c401a0cc8178b8, please let me know if you think this works. I decided to attach this to the previous paragraph, because it kind of belongs to it; please let me know if you like that.

Reviewer 2: This reads well to me. Thanks for clarifying, @blipp!

depend on the KEM. In the case of DHKEM, it is enough if the adversary
knows all private keys of one participant. 

## Domain Separation {#domain-separation}

HPKE allows combining a DHKEM variant DHKEM(Group, KDF') and a KDF
such that both KDFs are instantiated by the same KDF.

> 

Reviewer 2: ```suggestion
```

By design, the calls to Extract and Expand inside DHKEM have different prefix-free

> prefixes for their second parameter than the calls to Extract and

Reviewer 2: ```suggestion
encodings for their second parameter than the calls to Extract and
```

Expand in the remainder of HPKE. This is ensured by the different
prefix-free label parameters in the calls to LabeledExtract and
LabeledExpand. This serves to separate their input domains and justifies
modeling them as independent functions even if instantiated by the same

> KDF.

Reviewer 2: ```suggestion
By design, the calls to Extract and Expand inside DHKEM and the remainder of HPKE 
have different prefix-free encodings for the second parameter. This is achieved by the 
different prefix-free label parameters in the calls to LabeledExtract and LabeledExpand. 
This serves to separate all Extract and Expand invocation input domains. It also justifies 
modeling them as independent functions even if instantiated by the same KDF.
```

Future KEM instantiations MUST ensure that all internal invocations of
Extract and Expand can be modeled as functions independent from the
invocations of Extract and Expand in the remainder of HPKE. One way to
ensure this is by using an equal or similar prefixing scheme with
an identifier different from "RFCXXXX ". Particular attention needs to
be paid if the KEM directly invokes a function that is used internally

> in HPKE's Extract or Expand, like Hash in case of HKDF. It MUST be

Reviewer 2: ```suggestion
in HPKE's Extract or Expand, such as Hash in case of HKDF. It MUST be
```

ensured that inputs to these invokation cannot collide with inputs used

> inside Extract or Expand.

Reviewer 2: ```suggestion
in HPKE's Extract or Expand, such as Hash in the case of HKDF. Inputs to these
invocations MUST apply domain separation such that they cannot collide
with inputs used inside Extract or Expand.
```

Author: (I don't know why but it's marked as outdated for me, already before my most recent commit)
I applied the suggestions with the “such” and “the”. Domain separation is not the solution here because we cannot change the inputs to the internal invocations.

> The string literal "RFCXXX" used in LabeledExtract and LabeledExpand

Reviewer 2: ```suggestion
The string literal "RFCXXXX" used in LabeledExtract and LabeledExpand
```
